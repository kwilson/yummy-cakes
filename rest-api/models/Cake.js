class Cake {
    constructor(id, name, imageUrl, yumFactor, comment) {
        this.id = id;
        this.name = name;
        this.comment = comment;
        this.imageUrl = imageUrl;
        this.yumFactor = yumFactor;
    }
}

module.exports = Cake;