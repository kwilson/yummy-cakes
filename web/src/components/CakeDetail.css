.cake-detail {
    color: white;
    position: absolute;
    background: rgba(0,0,0,.8);
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    padding: 40px 20px;
    text-align: center;
    overflow: auto;
}

.cake-detail__img-container {
    text-align: center;
}

.cake-detail__img {
    height: auto;
    max-width: 100%;
}


.cake-detail__comment blockquote {
    font-style: italic;
}

    .cake-detail__comment blockquote::before,
    .cake-detail__comment blockquote::after {
        content: '"';
    }

    .cake-detail__close {
        background: rgb(169, 6, 6);
        border: solid 1px white;
        padding: 5px;
        cursor: pointer;
        color: white;
        position: absolute;
        top: 10px;
        right: 10px;
        text-transform: uppercase;
    }
