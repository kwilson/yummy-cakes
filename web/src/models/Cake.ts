export interface Cake {
  id: string;
  name: string;
  imageUrl: string;
  yumFactor: number;
  comment: string;
}
