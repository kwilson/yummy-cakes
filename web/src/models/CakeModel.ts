export interface CakeModel {
  id: string;
  name?: string;
  imageUrl?: string;
  yumFactor?: number;
  comment?: string;
}
