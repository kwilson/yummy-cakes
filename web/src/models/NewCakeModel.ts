export interface NewCakeModel {
  name?: string;
  imageUrl?: string;
  yumFactor?: number;
  comment?: string;
}
