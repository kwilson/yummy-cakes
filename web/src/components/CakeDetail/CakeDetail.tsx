import * as React from 'react';
import './CakeDetail.css';
import { CakeModel } from '../../models/CakeModel';

import { YumFactor } from '../';

export interface CakeDetailStateProps {
  selectedCake?: CakeModel;
  close: () => any;
}

const CakeDetail: React.StatelessComponent<CakeDetailStateProps> = ({ selectedCake, close }) => {
  if (!selectedCake) {
    return null;
  }

  const comment = selectedCake.comment
    ? <blockquote>{selectedCake.comment}</blockquote>
    : 'No comments';

  return (
    <div className="cake-detail">
      <div className="cake-detail__img-container">
        <img className="cake-detail__img" src={selectedCake.imageUrl} />
      </div>

      <h1 className="cake-detail__heading">{selectedCake.name}</h1>

      <div className="cake-detail__yum">
        <YumFactor value={selectedCake.yumFactor} />
      </div>

      <div className="cake-detail__comment">
        <h2>Comments</h2>

        <div className="cake-detail__comment-value">
          {comment}
        </div>
      </div>

      <button onClick={close} className="cake-detail__close">
        Close
      </button>
    </div>
  );
};

export default CakeDetail;
