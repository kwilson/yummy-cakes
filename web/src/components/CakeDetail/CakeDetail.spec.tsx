import * as React from 'react';
import { shallow } from 'enzyme';
import { create as createSnapshot } from 'react-test-renderer';

import CakeDetail from './CakeDetail';
import { CakeModel } from '../../models/CakeModel';

import { YumFactor } from '../';

describe('CakeDetail Component', () => {
  let closeHandler: jest.Mock<{}>;
  beforeEach(() => {
    closeHandler = jest.fn();
  });

  it('renders null if no cake is selected', () => {
    // Arrange
    const cmp = <CakeDetail close={closeHandler} />;

    // Act
    const el = shallow(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    // Assert
    expect(el.html()).toBe(null);
    expect(snapshot).toMatchSnapshot();
  });

  it('renders detals of the cake', () => {
    // Arrange
    const name = 'Name Value';
    const imageUrl = 'http://imageurl.invalid';
    const yumFactor = 4;
    const comment = 'Comment value';

    const cake: CakeModel = {
      id: 'ID123',
      name,
      imageUrl,
      yumFactor,
      comment
    };

    const cmp = <CakeDetail selectedCake={cake} close={closeHandler} />;

    // Act
    const el = shallow(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    expect(el.find('.cake-detail__img').props().src).toEqual(imageUrl);
    expect(el.find('.cake-detail__heading').text()).toEqual(name);
    expect(el.find('.cake-detail__comment-value').text()).toEqual(comment);

    expect(el.find(YumFactor).props().value).toEqual(yumFactor);

    expect(snapshot).toMatchSnapshot();
  });

  it('renders empty comment text if there are no comments', () => {
    // Arrange
    const name = 'Name Value';

    const cake: CakeModel = {
      id: 'ID123',
      name
    };

    const cmp = <CakeDetail selectedCake={cake} close={closeHandler} />;

    // Act
    const el = shallow(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    expect(el.find('.cake-detail__comment-value').text()).toEqual('No comments');

    expect(snapshot).toMatchSnapshot();
  });

  it('fires the close handler on button click', () => {
    // Arrange
    const name = 'Name Value';

    const cake: CakeModel = {
      id: 'ID123',
      name
    };

    const cmp = <CakeDetail selectedCake={cake} close={closeHandler} />;
    const el = shallow(cmp);

    // Act
    el.find('.cake-detail__close').simulate('click');

    // Assert
    expect(closeHandler).toHaveBeenCalledTimes(1);
  });
});
