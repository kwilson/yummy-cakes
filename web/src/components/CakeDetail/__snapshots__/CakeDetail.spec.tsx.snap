// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CakeDetail Component renders detals of the cake 1`] = `
<div
  className="cake-detail"
>
  <div
    className="cake-detail__img-container"
  >
    <img
      className="cake-detail__img"
      src="http://imageurl.invalid"
    />
  </div>
  <h1
    className="cake-detail__heading"
  >
    Name Value
  </h1>
  <div
    className="cake-detail__yum"
  >
    <div
      className="yum-factor"
    >
      Yum Factor
      <div
        aria-label="4 out of 5"
        className="yum-factor__rating"
        title="4 out of 5"
      >
        ★★★★☆
      </div>
    </div>
  </div>
  <div
    className="cake-detail__comment"
  >
    <h2>
      Comments
    </h2>
    <div
      className="cake-detail__comment-value"
    >
      <blockquote>
        Comment value
      </blockquote>
    </div>
  </div>
  <button
    className="cake-detail__close"
    onClick={[Function]}
  >
    Close
  </button>
</div>
`;

exports[`CakeDetail Component renders empty comment text if there are no comments 1`] = `
<div
  className="cake-detail"
>
  <div
    className="cake-detail__img-container"
  >
    <img
      className="cake-detail__img"
      src={undefined}
    />
  </div>
  <h1
    className="cake-detail__heading"
  >
    Name Value
  </h1>
  <div
    className="cake-detail__yum"
  />
  <div
    className="cake-detail__comment"
  >
    <h2>
      Comments
    </h2>
    <div
      className="cake-detail__comment-value"
    >
      No comments
    </div>
  </div>
  <button
    className="cake-detail__close"
    onClick={[Function]}
  >
    Close
  </button>
</div>
`;

exports[`CakeDetail Component renders null if no cake is selected 1`] = `null`;
